lib/tlIO/FFmpegWrite.cpp

```diff
@@ -42,198 +176,582 @@ namespace tl
+                        resample = true;
 
-            p.avVideoStream->time_base = { rational.second, rational.first };
-            p.avVideoStream->avg_frame_rate = { rational.first, rational.second };
+                        if (checkSampleFormat(avCodec, AV_SAMPLE_FMT_FLT))
+                        {
+                            p.avAudioPlanar = false;
+                            p.avAudioCodecContext->sample_fmt =
+                                AV_SAMPLE_FMT_FLT;
+                        }
+                        else if(checkSampleFormat(avCodec, AV_SAMPLE_FMT_FLTP))
+                        {
+                            p.avAudioPlanar = true;
+                            p.avAudioCodecContext->sample_fmt =
+                                AV_SAMPLE_FMT_FLTP;
+                        }
+                        else if(checkSampleFormat(avCodec, AV_SAMPLE_FMT_S16))
+                        {
+                            p.avAudioPlanar = false;
+                            p.avAudioCodecContext->sample_fmt =
+                                AV_SAMPLE_FMT_S16;
+                        }
+                        else
+                        {
+                            throw std::runtime_error(
+                                string::Format(
+                                    "Sample format {0} not supported!")
+                                    .arg(av_get_sample_fmt_name(
+                                        p.avAudioCodecContext->sample_fmt)));
+                        }
+                    }
+                    else
+                    {
+                        p.avAudioCodecContext->sample_fmt = planarFormat;
+                        p.avAudioPlanar = true;
+                    }
+                }
 
-            for (const auto& i : info.tags)
-            {
-                av_dict_set(&p.avFormatContext->metadata, i.first.c_str(), i.second.c_str(), 0);
-            }
+                if (p.avAudioPlanar)
+                    p.flatData.resize(info.audio.channelCount);
+                else
+                    p.flatData.resize(1);
+                
+                r = selectChannelLayout(
+                    avCodec, &p.avAudioCodecContext->ch_layout,
+                    info.audio.channelCount);
+                if (r < 0)
+                    throw std::runtime_error(
+                        string::Format(
+                            "{0}: Could not select audio channel layout")
+                            .arg(p.fileName));
 
-            //av_dump_format(p.avFormatContext, 0, p.fileName.c_str(), 1);
+                p.sampleRate =
+                    selectSampleRate(avCodec, info.audio.sampleRate);
+                if (p.sampleRate == 0)
+                    throw std::runtime_error(
+                        string::Format(
+                            "{0}: Could not select sample rate")
+                            .arg(p.fileName));
+                    
+                char buf[256];
```

Comment:
It would be nice to use a std::vector instead of C style array here. If the size needs to change later it makes it easier to update:
```
std::vector<char> buf(256);
av_channel_layout_describe(&p.avAudioCodecContext->ch_layout, buf.data(), buf.size());
```